RecoBTag/TensorFlow/plugins/DeepVertexTFJetTagsProducer.cc

```diff
@@ -0,0 +1,323 @@
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/Framework/interface/makeRefToBaseProdFrom.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/BTauReco/interface/JetTag.h"
+
+#include "DataFormats/BTauReco/interface/DeepFlavourTagInfo.h"
+
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+
+#include "RecoBTag/TensorFlow/interface/tensor_fillers.h"
+
+
+// Declaration of the data structure that is hold by the edm::GlobalCache.
+// In TensorFlow, the computational graph is stored in a stateless graph object which can be shared
+// by multiple session instances which handle the initialization of variables related to the graph.
+// Following this approach in CMSSW, a graph should be stored in a GlobalCache which can be accessed
+// by sessions owned by multiple stream module copies. Instead of using only the plain graph, we
+// make use of a cache struct that can be extended in the future if nedded. In addition, the graph
+// is protected via std::atomic, which should not affect the performance as it is only accessed in
+// the module constructor and not in the actual produce loop.
+struct DeepVertexTFCache {
+  DeepVertexTFCache() : graphDef(nullptr) {
+  }
+
+  std::atomic<tensorflow::GraphDef*> graphDef;
+};
+
+
+class DeepVertexTFJetTagsProducer : public edm::stream::EDProducer<edm::GlobalCache<DeepVertexTFCache>> {
+
+  public:
+    explicit DeepVertexTFJetTagsProducer(const edm::ParameterSet&, const DeepVertexTFCache*);
+    ~DeepVertexTFJetTagsProducer() override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+    static std::unique_ptr<DeepVertexTFCache> initializeGlobalCache(const edm::ParameterSet&);
+    static void globalEndJob(const DeepVertexTFCache*);
+
+    enum InputIndexes {
+      kGlobal = 0,
+      kSeedingTracks = 1,
+      kNeighbourTracks = 2,
+      
+    };
+
+    enum OutputIndexes {
+      kJetFlavour = 0,     
+        
+    };
+    
+  private:
+    typedef std::vector<reco::DeepFlavourTagInfo> TagInfoCollection;
+    typedef reco::JetTagCollection JetTagCollection;
+
+    void beginStream(edm::StreamID) override {}
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override {}
+
+    const edm::EDGetTokenT< TagInfoCollection > src_;
+    std::vector<std::pair<std::string,std::vector<unsigned int>>> flav_pairs_;
+    std::vector<std::string> input_names_;
+    std::vector<std::string> output_names_;
+    std::vector<std::string> lp_names_;
+
+    // session for TF evaluation
+    tensorflow::Session* session_;
+    // vector of learning phase tensors, i.e., boolean scalar tensors pointing to false
+    std::vector<tensorflow::Tensor> lp_tensors_;
+    // flag to evaluate model batch or jet by jet
+    bool batch_eval_;
+    
+};
+
+
+DeepVertexTFJetTagsProducer::DeepVertexTFJetTagsProducer(const edm::ParameterSet& iConfig,
```

Comment:
@slava77 this could be done at some point but we would prefer not to entangle the two developements right now as one is in production and the other is still in development. As we plan to commit also a merged architecture DeepVertex + DeepFlavour at that point we may propose a common base class with three concrete implementations